WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+multiple branches that go back to different script executions. This is
+particularly important when the data flow in implementations is expected to be
+different from the spec in relevant ways.
+
+An example of this is HTML media playback events, where if the user clicks play,
+and in short succession JS code calls `videoEl.load()` and then `videoEl.play()`
+in two different contexts, all three data flow branches will merge, resulting in
+a single `load` event being fired.
+
+If some of those sources are browser-originated and some originate from JS, only
+the latter ones should be considered. Beyond that, each such individual event
+would have to be considered. For media playback events, this merge is caused by
+debouncing (i.e. if a load is already in progress, calling a method that would
+start one will reuse the existing one), and so the dispatch context should be
+that for the earliest web API call that resulted in this event. But other cases
+might have other needs, and their specifics need to be considered.
+
+## Runtime script errors
+
+The `error` event on a window or worker global object is fired whenever a script
+execution throws an uncaught exception. Having access to the context in which
+the exception was thrown can help determine which of multiple independent
+streams of execution triggered this error, and therefore how to clean up after
+it. Therefore,
+[`ErrorEvent`](https://html.spec.whatwg.org/multipage/webappapis.html#errorevent)
+will have a `throwSnapshot` property, reflecting the context in which the
+exception was thrown.
+
+
+## Promises / async web APIs
+
+The `unhandledrejection` event on a window or worker global object is fired
+whenever a promise resolves with a rejection, without having a rejection handler
+registered. Exposing the context in which the promise was rejected is useful for
+the same reasons as for runtime script errors. Therefore,
+[`PromiseRejectionEvent`](https://html.spec.whatwg.org/multipage/webappapis.html#promiserejectionevent)
+will have a `rejectionSnapshot` property reflecting this context.
+
+In the case of promises created by web APIs, it seems the rejection context
+would always be the same as the context when the promise is created. According
+to the categories in the [“Writing Promise-Using
+Specifications”](https://w3ctag.github.io/promises-guide/) guide:
+- For one-and-done operations, the rejection-time context of the returned
+  promise should be the context when the web API that returns it was called.
+- For one-time “events”, the promise is created at the same time as an async
+  operation is started which will eventually resolve it, so the context would
+  flow from creation to rejection. E.g, for the
+  [`loaded`](https://drafts.csswg.org/css-font-loading-3/#dom-fontface-loaded)
+  property of a [`FontFace`](https://drafts.csswg.org/css-font-loading-3/#fontface)
+  instance, creating the `FontFace` instance causes both the promise creation
+  and the loading of the font.
```

Comment:
This doesn't seem accurate. E.g. a stream could be created in one context, and then errored (which causes `ready` to reject) in another.